smithy-codegen-core/src/test/java/software/amazon/smithy/codegen/core/IntegrationTopologicalSortTest.java

```diff
@@ -0,0 +1,218 @@
+        return new SmithyIntegration.Order() {
+            @Override
+            public byte priority() {
+                return priority;
+            }
+
+            @Override
+            public List<String> runBefore() {
+                return runBefore == null ? Collections.emptyList() : ListUtils.of(runBefore);
+            }
+
+            @Override
+            public List<String> runAfter() {
+                return runAfter == null ? Collections.emptyList() : ListUtils.of(runAfter);
+            }
+        };
+    }
+
+    static List<String> toStrings(List<? extends SmithyIntegration<MySettings>> integrations) {
+        return SmithyIntegration.sort(integrations).stream()
+                .map(SmithyIntegration::name)
+                .collect(Collectors.toList());
+    }
+
+    @Test
+    public void sortsIntegrationsWithAllDefaults() {
+        List<MyIntegration> integrations = new ArrayList<>();
+        integrations.add(new MyIntegration("a"));
+        integrations.add(new MyIntegration("b"));
+        integrations.add(new MyIntegration("c"));
+        integrations.add(new MyIntegration("d"));
+
+        List<String> result = toStrings(SmithyIntegration.sort(integrations));
+
+        // Note that insertion order is maintained.
+        assertThat(result, contains("a", "b", "c", "d"));
+    }
+
+    @Test
+    public void sortsIntegrationsBasedOnlyOnPriority1() {
+        List<MyIntegration> integrations = new ArrayList<>();
+        integrations.add(new MyIntegration("a", createOrder((byte) 0)));
+        integrations.add(new MyIntegration("b", createOrder((byte) 1)));
+        integrations.add(new MyIntegration("c", createOrder((byte) 2)));
+        integrations.add(new MyIntegration("d", createOrder((byte) 3)));
+
+        List<String> result = toStrings(SmithyIntegration.sort(integrations));
+
+        assertThat(result, contains("d", "c", "b", "a"));
+    }
+
+    @Test
+    public void sortsIntegrationsBasedOnlyOnPriority2() {
+        List<MyIntegration> integrations = new ArrayList<>();
+        integrations.add(new MyIntegration("a", createOrder((byte) 2)));
+        integrations.add(new MyIntegration("b", createOrder((byte) 0)));
+        integrations.add(new MyIntegration("c", createOrder((byte) 1)));
+        integrations.add(new MyIntegration("d", createOrder((byte) 3)));
+
+        List<String> result = toStrings(SmithyIntegration.sort(integrations));
+
+        assertThat(result, contains("d", "a", "c", "b"));
+    }
+
+    @Test
+    public void sortsIntegrationsBasedOnBefore() {
+        List<MyIntegration> integrations = new ArrayList<>();
+        integrations.add(new MyIntegration("a", createOrder((byte) 0, "b")));
+        integrations.add(new MyIntegration("b", createOrder((byte) 0, "c")));
+        integrations.add(new MyIntegration("c", createOrder((byte) 0, "d")));
+        integrations.add(new MyIntegration("d", createOrder((byte) 0)));
```

Comment:
You should shuffle this around so the expected result isn't also insertion order

Reply:
There are other tests that do out of order checks too